File: examples/python/reaction_role.py
"""反应角色最小示例。"""

from __future__ import annotations

import asyncio
import os

from newtspeak_bot import OwlBotClient, run_gateway

bot = OwlBotClient(os.environ["NEWT_BASE_URL"], token=os.environ["NEWT_BOT_TOKEN"])
RULES_MESSAGE_ID = os.environ["RULES_MESSAGE_ID"]
VERIFIED_ROLE_ID = os.environ["VERIFIED_ROLE_ID"]


async def on_event(event: str, data: dict) -> None:
    if event == "READY":
        print("已连接", (data or {}).get("user", {}).get("username"))
        return
    if event != "MESSAGE_REACTION_ADD":
        return
    if str(data.get("message_id")) != RULES_MESSAGE_ID or data.get("emoji") != "✅":
        return
    try:
        bot.add_member_role(data["guild_id"], data["user_id"], VERIFIED_ROLE_ID)
        print("已赋角", data["user_id"])
    except Exception as exc:  # noqa: BLE001
        print("赋角失败", exc)


if __name__ == "__main__":
    asyncio.run(run_gateway(bot, on_event))

File: examples/rust/reaction_role.rs
//! 反应角色最小示例（作为文档参考；可拷贝到独立 binary crate）。
//!
//! 运行需自行建 package 并依赖 path = "../../rust"。

use newtspeak_bot::{connect_gateway, Client, HandlerMap};
use std::sync::Arc;

#[tokio::main]
async fn main() -> newtspeak_bot::Result<()> {
    let bot = Client::new(
        std::env::var("NEWT_BASE_URL").expect("NEWT_BASE_URL"),
        std::env::var("NEWT_BOT_TOKEN").expect("NEWT_BOT_TOKEN"),
    )?;
    let rules = Arc::new(std::env::var("RULES_MESSAGE_ID").expect("RULES_MESSAGE_ID"));
    let role = Arc::new(std::env::var("VERIFIED_ROLE_ID").expect("VERIFIED_ROLE_ID"));

    let handlers = HandlerMap::new();
    let bot_c = bot.clone();
    let rules_c = rules.clone();
    let role_c = role.clone();
    handlers
        .on("MESSAGE_REACTION_ADD", move |_, ev| {
            let bot = bot_c.clone();
            let rules = rules_c.clone();
            let role = role_c.clone();
            tokio::spawn(async move {
                let mid = ev.get("message_id").map(|v| v.to_string()).unwrap_or_default();
                let emoji = ev.get("emoji").and_then(|v| v.as_str()).unwrap_or("");
                if mid.trim_matches('"') != rules.as_str() || emoji != "✅" {
                    return;
                }
                let guild = ev.get("guild_id").and_then(|v| v.as_str()).unwrap_or("");
                let user = ev.get("user_id").and_then(|v| v.as_str()).unwrap_or("");
                if let Err(e) = bot.add_member_role(guild, user, role.as_str()).await {
                    eprintln!("赋角失败: {e}");
                } else {
                    println!("已赋角 {user}");
                }
            });
        })
        .await;

    let _gw = connect_gateway(bot, handlers).await?;
    tokio::signal::ctrl_c().await.ok();
    Ok(())
}
